File: memory/split.py

import base
import parser

def random_split(machine, nxt, rand, cost):
   offset = rand.random_address(machine.word_size)
   bank0 = base.Join(0)
   bank1 = base.Join(1)
   result = Split(bank0, bank1, nxt, offset)
   if result.get_cost() <= cost:
      return result
   else:
      return None

class Split(base.Memory):

   def __init__(self, bank0, bank1, mem, offset):
      self.bank0 = bank0
      self.bank1 = bank1
      self.mem = mem
      self.offset = offset
      base.set_parent(bank0, self)
      base.set_parent(bank1, self)

   def __str__(self):
      result  = "(split "
      result += "(offset " + str(self.offset) + ")"
      result += "(bank0 " + str(self.bank0) + ")"
      result += "(bank1 " + str(self.bank1) + ")"
      result += "(memory " + str(self.mem) + ")"
      result += ")"
      return result

   def get_next(self):
      return self.mem

   def set_next(self, n):
      self.mem = n

   def get_banks(self):
      return [self.bank0, self.bank1]

   def set_bank(self, i, b):
      assert(i >= 0 and i <= 1)
      if i == 0:
         self.bank0 = b
      else:
         self.bank1 = b

   def permute(self, rand, max_cost):
      self.offset = rand.random_address(self.machine.word_size)
      return True

   def simplify(self):
      self.bank0 = self.bank0.simplify()
      self.bank1 = self.bank1.simplify()
      self.mem = self.mem.simplify()
      if isinstance(self.bank0, base.Join) and \
         isinstance(self.bank1, base.Join):
         return self.mem
      return self

   def done(self):
      return self.mem.done()

   def process(self, start, write, addr, size):
      mask = self.machine.addr_mask
      last = (addr + size - 1) & mask
      result = start
      if addr > last:
         result = self._do_process(result, addr, mask - addr + 1, write)
         result = self._do_process(result, 0, last + 1, write)
      else:
         result = self._do_process(result, addr, size, write)
      return result

   def _do_process(self, start, addr, size, write):
      last = (addr + size - 1) & self.machine.addr_mask
      result = start
      if addr < self.offset:
         if last <= self.offset:
            temp_size = size
         else:
            temp_size = self.offset - addr
         result = self.bank0.process(result, write, addr, temp_size)
      if last >= self.offset:
         if addr >= self.offset:
            temp_addr = addr - self.offset
            temp_size = size
         else:
            temp_addr = 0
            temp_size = last - self.offset + 1
         result = self.bank1.process(result, write, temp_addr, temp_size)
      return result

   def forward(self, index, start, write, addr, size):
      if index == 1:
         addr = (addr + self.offset) & self.machine.addr_mask
         return self.mem.process(start, write, addr, size)
      else:
         return self.mem.process(start, write, addr, size)

def _create_split(lexer, args):
   offset = parser.get_argument(lexer, args, 'offset', 0)
   mem = parser.get_argument(lexer, args, 'memory')
   bank0 = parser.get_argument(lexer, args, 'bank0')
   bank1 = parser.get_argument(lexer, args, 'bank1')
   return Split(bank0, bank1, mem, offset)
base.constructors['split'] = _create_split

File: fastsim/split.ml
open Machine
open Base_memory

class split =
    object (self)
        inherit container as super

        val mutable offset : int = 0
        val mutable bank0 : base_memory option = None
        val mutable bank1 : base_memory option = None

        method set name value =
            match name with
            | "offset" -> offset <- int_of_string value
            | _ -> super#set name value

        method set_bank0 b =
            bank0 <- b;
            self#bank0#set_parent (self :> base_memory)

        method set_bank1 b =
            bank1 <- b;
            self#bank1#set_parent (self :> base_memory)

        method bank0 =
            match bank0 with
            | Some b -> b
            | None -> failwith "invalid bank0"

        method bank1 =
            match bank1 with
            | Some b -> b
            | None -> failwith "invalid bank1"

        method reset m main =
            super#reset m main;
            self#bank0#reset m main;
            self#bank1#reset m main

        method private process base start write addr size =
            let mask = mach.addr_mask in
            let last = (addr + size - 1) land mask in
            let result = start in
            if addr > last then
                let first_size = mask - addr + 1 in
                let t = self#do_process base result addr first_size write in
                self#do_process base t 0 (last + 1) write
            else
                self#do_process base result addr size write

        method private do_process base start addr size write =
            let last = (addr + size - 1) land mach.addr_mask in
            let start =
                if addr < offset then
                    let temp_size =
                        if last < offset then size
                        else offset - addr in
                    self#bank0#send_request base start write addr temp_size
                else start
            in
            if last >= offset then
                let temp_addr, temp_size = 
                    if addr >= offset then (addr - offset, size)
                    else (0, last - offset + 1) in
                self#bank1#send_request base start write temp_addr temp_size
            else start

        method forward base index start write addr size =
            let addr =
                match index with
                | 0 -> addr
                | 1 -> (addr + offset) land mach.addr_mask
                | _ -> failwith "invalid"
            in self#next#send_request base start write addr size

    end
